com.unity.ml-agents/Runtime/Inference/TensorGenerator.cs

```diff
@@ -94,30 +94,39 @@ public interface IGenerator
         public void InitializeObservations(List<ISensor> sensors, ITensorAllocator allocator)
         {
             // Loop through the sensors on a representative agent.
-            // For vector observations, add the index to the (single) VectorObservationGenerator
-            // For visual observations, make a VisualObservationInputGenerator
+            // For vector observations, add the index to the (single) ObservationGenerator
+            // For visual observations, make a NonVectorObservationInputGenerator
             var visIndex = 0;
-            VectorObservationGenerator vecObsGen = null;
+            ObservationGenerator vecObsGen = null;
             for (var sensorIndex = 0; sensorIndex < sensors.Count; sensorIndex++)
             {
                 var sensor = sensors[sensorIndex];
                 var shape = sensor.GetObservationShape();
                 // TODO generalize - we currently only have vector or visual, but can't handle "2D" observations
-                var isVectorSensor = (shape.Length == 1);
-                if (isVectorSensor)
+                var rank = shape.Length;
+                switch (rank)
```

Comment:
You can DRY up this a bit with something like
```csharp
                ObservationGenerator obsGen = null;
                string obsGenName = null;
                switch (rank)
                {
                    case 1:
                        if (vecObsGen == null)
                        {
                            vecObsGen = new ObservationGenerator(allocator);
                        }
                        obsGen = vecObsGen;
                        obsGenName = TensorNames.VectorObservationPlaceholder;
                        break;
                    case 2:
                        obsGen = new ObservationGenerator(allocator);
                        obsGenName = TensorNames.ObservationPlaceholderPrefix + sensorIndex;
                        break;
                    case 3:
                        obsGen = new ObservationGenerator(allocator);
                        obsGenName = TensorNames.VisualObservationPlaceholderPrefix + visIndex;
                        visIndex++;
                        break;
                    default:
                        // TODO error handling
                        break;
                }

                obsGen.AddSensorIndex(sensorIndex);
                m_Dict[obsGenName] = obsGen;
```
and then delete this block below:
```csharp
            if (vecObsGen != null)
            {
                m_Dict[TensorNames.VectorObservationPlaceholder] = vecObsGen;
            }
```